service.go

```diff
@@ -0,0 +1,194 @@
+package sync
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type DefaultService struct {
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	barriersMu sync.Mutex
+	barriers   map[string]*barrier
+	subsMu     sync.Mutex
+	subs       map[string]*pubsub
+}
+
+func NewDefaultService(ctx context.Context, log *zap.SugaredLogger) (*DefaultService, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	s := &DefaultService{
+		ctx:      ctx,
+		cancel:   cancel,
+		barriers: map[string]*barrier{},
+		subs:     map[string]*pubsub{},
+	}
+
+	s.wg.Add(2)
+	go s.subscriptionGC()
+	go s.barriersGC()
+
+	return s, nil
+}
+
+func (s *DefaultService) Publish(ctx context.Context, topic string, payload interface{}) (int, error) {
+	log := log.With("topic", topic)
+	log.Debugw("publishing item on topic", "payload", payload)
+
+	bytes, err := json.Marshal(payload)
+	if err != nil {
+		return -1, fmt.Errorf("failed while serializing payload: %w", err)
+	}
+
+	log.Debugw("serialized json payload", "json", string(bytes))
+
+	s.subsMu.Lock()
+	defer s.subsMu.Unlock()
+
+	ps := s.createSubIfNew(topic)
+	seq := ps.publish(string(bytes))
+
+	log.Debugw("successfully published item; sequence number obtained", "seq", seq)
+	return seq, nil
+}
+
+func (s *DefaultService) Subscribe(ctx context.Context, topic string) (*subscription, error) {
+	s.subsMu.Lock()
+	defer s.subsMu.Unlock()
+
+	ps := s.createSubIfNew(topic)
+	return ps.subscribe(ctx), nil
+}
+
+func (s *DefaultService) createSubIfNew(topic string) *pubsub {
+	if _, ok := s.subs[topic]; !ok {
+		ctx, cancel := context.WithCancel(s.ctx)
+		s.subs[topic] = &pubsub{
+			ctx:     ctx,
+			cancel:  cancel,
+			lastmod: time.Now(),
+			msgs:    []string{},
+			subs:    []*subscription{},
+		}
+		go s.subs[topic].worker()
+	}
+
+	return s.subs[topic]
+}
+
+func (s *DefaultService) Barrier(ctx context.Context, state string, target int) error {
```

Comment:
I think it'd be good to have debug logging here as well, same as with the other public methods.